onnxruntime/core/providers/cuda/cuda_graph.cc

```diff
@@ -0,0 +1,89 @@
+#include "core/providers/cuda/cuda_graph.h"
+
+#include "core/providers/cuda/cuda_common.h"
+#include <cuda_runtime_api.h>
+#include <driver_types.h>
+
+
+namespace onnxruntime {
+
+CUDAGraph::CUDAGraph(cudaStream_t stream) : capture_stream_(stream) {
+#if (defined(CUDA_VERSION) && CUDA_VERSION < 10000)
+  ORT_THROW("CUDA graphs can only be used in Onnxruntime built with CUDA >= 11.0");
+#endif
+}
+
+void CUDAGraph::CaptureBegin() {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 10000
+  ORT_ENFORCE(!has_graph_exec_,
+              "This cuda graph has already captured a graph. "
+              "Create a new instance to capture a new graph.");
+
+  CUDA_CALL_THROW(cudaDeviceSynchronize());
+  CUDA_CALL_THROW(cudaStreamBeginCapture(capture_stream_, cudaStreamCaptureModeGlobal));
+#else
+  ORT_THROW("CUDA graphs can only be used in Onnxruntime built with CUDA >= 11.0");
+#endif
+}
+
+void CUDAGraph::CaptureEnd() {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 10000
+  CUDA_CALL_THROW(cudaStreamEndCapture(capture_stream_, &graph_));
+  if (graph_ == NULL) {
+    ORT_THROW("CUDAGraph::CaptureEnd: graph_ is NULL");
+  }
+
+  has_graph_ = true;
+  CUDA_CALL_THROW(cudaGraphInstantiate(&graph_exec_, graph_, NULL, NULL, 0));
+  has_graph_exec_ = true;
+  CUDA_CALL_THROW(cudaGraphDestroy(graph_));
+  has_graph_ = false;
+#else
+  ORT_THROW("CUDA graphs can only be used in Onnxruntime built with CUDA >= 11.0");
+#endif
+}
+
+Status CUDAGraph::Replay() {
+  std::lock_guard<OrtMutex> lock(lock_);
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 10000
+  CUDA_RETURN_IF_ERROR(cudaGraphLaunch(graph_exec_, capture_stream_));
+
+  int version;
+  CUDA_RETURN_IF_ERROR(cudaDriverGetVersion(&version));
+  if (version < 11040) {
+    // Workaround for bug in libcuda.so that causes replayed graphs with
+    // certain topologies to be corrupted (kernels elided, internal syncs
+    // ignored) when replayed back to back without a sync in between.
+    // The bug is fixed in CUDA 11.4+.
+    CUDA_RETURN_IF_ERROR(cudaDeviceSynchronize());
```

Comment:
Do we need a costly device sync ? Or will a stream sync do ? If a stream sync will do, don't we do it anyway in OnRFunEnd() ? If so, do we need this here ?

Reply:
This is for a bug in CUDA <= 11.4.  The code was borrowed from pytorch [here](https://github.com/pytorch/pytorch/blob/5cae40c169b3da48d2782947d072d20b7acb15d1/aten/src/ATen/cuda/CUDAGraph.cpp#L183-L191). The related PRs are [here](https://github.com/pytorch/pytorch/pull/61063) and [here](https://github.com/pytorch/pytorch/pull/57556). PyTorch solved this issue by using device sync instead of a stream sync. I'm not quite sure about the root cause. I'm guessing there may be some reasons behind it.

Reply:
Based on offline discussion, this was removed and we will add it if the users experience issues around this.